### ISSUE TITLE
[intro.races] p16 Can this rule infer the modification order according to the reading value?

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#622]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[intro.races] p16 says;
> If a value computation A of an atomic object M happens before a value computation B of M, and A takes its value from a side effect X on M, then the value computed by B is either the value stored by X or the value stored by a side effect Y on M, where Y follows X in the modification order of M.

I thought this rule is for this example:
````text
// thread 1:
x.store(1);  // #1
x.store(2); // #2
// thread 2:
x.load();  // #3
x.load(); // #4
````
According to [intro.races] p15, `#2` follows `#1` in the modification order of `x`, if `#3` reads `1`, then `#4` may read `1` or `2`. In this example, the sentence `where Y follows X in the modification order of M` requires what `#4` can be. 

However, consider this example
````text
// thread 1:
x.store(3);  // #1
//thread 2:
x.store(4); // #2
//thread 3:
x.load(); // #3
x.load();  // #4 
````
if `#3` read `3` and `#4` read `4`, can we infer that `#2` follows `#1` in the modification order of `x` according to [intro.races] p16? 

### Suggested Resolution
In the first example, the `where` sentence impose the requirement, which means
>  B either reads the value stored by `X` or the value stored by `Y` that follows `X` in the modification order of `M`

while in the second example, the `where` sentence means
> If B reads the value stored by `Y`, then `Y` follows `X` in the modification order of `M`

In other words, the first interpretation describes we **impose the requirement** on the order of `Y` in the modification order, instead, the second interpretation **specifies the order of `Y` in the modification order. The rule intends to cover both interpretations, however, the `where` sentence is more inclined to the first interpretation.

**Comment by t3nsor** [comment]
p16 establishes a causal relationship that flows in one direction: the modification order imposes restrictions on the results of loads, not vice versa. Using logic, one can also reason backward from the observed behavior to the modification order if the program sees 3 and then 4. We don't need to change the wording in the standard just to remind readers about how propositional logic works.

**Comment by xmh0511** [comment]
@t3nsor 
> not vice versa

Taken
> a value computation A of an atomic object M happens before a value computation B of M, and A takes its value from a side effect X on M

as `p` and 
> the value computed by B is either the value stored by X or the value stored by a side effect Y on M, where Y follows X in the modification order of M

as `q`

did you mean `p -> q` does not mean `q->p`?

> how propositional logic works

When `p` is true, `p->q` is true if and only if `q` is true, in order not to violate the rule, the implementation must guarantee `q` is true, hence if `b` reads `y` that is not `x`, the consequence `q` is true when `y` satisfy the requirement `Y follows X in the modification order of M`

Is this how the propositional logic works you mentioned?

**Comment by t3nsor** [comment]
I suppose, to be precise, the standard says (P∧Q) -> R, where P is "a value computation A of an atomic object M happens before a value computation B of M", Q is "A takes its value from a side effect X on M", and R is "the value computed by B is either the value stored by X or the value stored by a side effect Y on M, where Y follows X in the modification order of M".

In your second example you observed that it is possible to infer (P∧¬R) -> ¬Q, which is equivalent to (P∧Q) -> R so the standard already allows for inferences in both directions but it is not necessary for the standard to explicitly state all rules that are logically equivalent to the ones that are already stated.

**Comment by xmh0511** [comment]
> In your second example you observed that it is possible to infer (P∧¬R) -> ¬Q, which is equivalent to (P∧Q) -> R so the standard already allows for inferences in both directions but it is not necessary for the standard to explicitly state all rules that are logically equivalent to the ones that are already stated.

From logic equivalent, `(P∧Q) -> R` equals to `¬R -> ¬P v ¬Q`, why do you say it is equivalent to `(P∧¬R) -> ¬Q`?

**Comment by t3nsor** [comment]
> > In your second example you observed that it is possible to infer (P∧¬R) -> ¬Q, which is equivalent to (P∧Q) -> R so the standard already allows for inferences in both directions but it is not necessary for the standard to explicitly state all rules that are logically equivalent to the ones that are already stated.
> 
> From logic equivalent, `(P∧Q) -> R` equals to `¬R -> ¬P v ¬Q`, why do you say it is equivalent to `(P∧¬R) -> ¬Q`?

Something can be equivalent to more than one thing

**Comment by jensmaurer** [comment]
I'm getting a bit frustrated with this discussion.

`A -> B` is defined to be equivalent to "not A or B".

Applying a bit of boolean logic yields:
`(P and Q) -> R` equals `not(P and Q) or R` equals `not P or not Q or R` equals `not P or R or not Q` equals `not (not P or R) -> not Q` equals `P and not R -> not Q`.
